source/opt/compact_ids_pass.cpp

```diff
@@ -44,6 +44,8 @@ Pass::Status CompactIdsPass::Process() {
   bool modified = false;
   std::unordered_map<uint32_t, uint32_t> result_id_mapping;
 
+  context()->DisableAnalyses(IRContext::kAnalysisDebugInfo);
```

Comment:
> Why can't you just invalidate the analysis?

This would "work" ie it would eliminate the assert my client's code is hitting.

But frankly, I do not like this solution.

I do like the InvalidateAnalyses() mechanism, but it was not designed to turn off an analysis for the duration of a pass, which is what is needed here.

The following code sequence around each AnalyzeDebugInst() does essentially disable the DebugInfoMgr for the duration of the pass in which it was invalidated, but I find it very klunky and unintuitive:

```
  if (!IsLineInst() &&
      context()->AreAnalysesValid(IRContext::kAnalysisDebugInfo)) {
    context()->get_debug_info_mgr()->AnalyzeDebugInst(this);
  }
```

I find it much more intuitive and clean to be able to explicitly disable and enable an analysis and query its status. In retrospect, I would probably change these to just:

```
  if (!IsLineInst() &&
      !context()->AreAnalysesDisabled(IRContext::kAnalysisDebugInfo)) {
    context()->get_debug_info_mgr()->AnalyzeDebugInst(this);
  }
```

If you disagree, I will cancel this PR and just submit a PR which does an InvalidateAnalyses(kAnalysisDebugInfo) at the top of CompactIds.

Reply:
Actually, I left in the check for AreAnalysesValid() in case there are still cases where it is needed.

Reply:
> The following code sequence around each AnalyzeDebugInst() does essentially disable the DebugInfoMgr for the duration of the pass in which it was invalidated, but I find it very klunky and unintuitive:

It does not "disable" the DebugInfoMgr.  Those lines say: If the debug info is valid, analyze the current instruction to keep it valid.  To make that clearer, we probably should have put that code into a function with a meaningful name.

> I find it much more intuitive and clean to be able to explicitly disable and enable an analysis and query its status. In retrospect, I would probably change these to just:

This is not cleaner.  If the analysis is not disabled, it might rebuild it from scratch when it is not needed.  The point of that line is that if the analysis is not valid, we do not want to build it at that time.  We only want to update it.  Your proposal changes the semantics of that code.

Also, if we use the concept of "disabling" a transformation in the context, you will need to add a lot more code.  Every place we potentially try to get an analysis, we will have to do something if it is disabled, like return null.  To handle this correctly, you will have to go to every place that tries to use an analysis check if it is disabled.  What if one pass disable it and then does not reenable it?  Does it become the responsibility of the pass writer to enable every analysis it might use?  The way you have it written now means you have to update every pass to make sure it enables all of the analysies. What do you do if an analysis is not disabled, but it depends on another one that is disabled?

I believe it is simpler for the user to have the following true at the start or end of every pass:

1) If an analysis is "valid", then it is in a state that would be equivalent to rebuilding it from scratch.
2) If  an analysis is not valid, then it can be rebuilt.

The responsibilities of the pass writer: as they change the code, it is their responsibility to not build or use an analysis at a point when it cannot be properly built.  The tricky part of this one is that the debug info is coming in already valid, and then it is automatically being updated when it should not be.

Also you will have to add testing in the ir context for it.  You will need to make sure that the ir context will never build an analysis if it is disable.

> If you disagree, I will cancel this PR and just submit a PR which does an InvalidateAnalyses(kAnalysisDebugInfo) at the top of CompactIds.

Yes, I disagree.  Please cancel this PR and open a new one where you just call InvalidateAnalyses.